mita/openstack.py

```diff
@@ -157,47 +157,73 @@ def _get_flavor(self) -> NodeSize:
             logger.error(be)
             raise OpenStackDriverError("Encountered an unknown exception.")
 
-    def _has_free_ip_address(self, network: str) -> bool:
+    def _get_network_by_name(self, name: str) -> OpenStackNetwork:
         """
-        Return True if the given network has an IPv4 Address to lease.
+        Retrieve the OpenStackNetwork instance using the provided name.
+
+        Args:
+            name:   the name of the network.
+
+        Returns:
+            OpenStackNetwork instance
+
+        Raises:
+            ResourceNotFound when no matching resource was found.
+        """
+        url = f"{self.driver_v2._networks_url_prefix}?name={name}"
+        object_ = self.driver_v2.network_connection.request(url).object
+        networks = self.driver_v2._to_networks(object_)
+
+        if not networks:
+            raise ResourceNotFound(f"No network resource with name {name} found.")
+
+        return networks[0]
+
+    def _has_free_ip_address(self, net: OpenStackNetwork) -> bool:
+        """
+        Return True if the given network has an ip address to lease.
 
         Arguments:
-            network: String, The network UUID
+            net:    The reference to OpenStackNetwork
 
         Returns:
             bool - True on success else False
 
         Note: This method returns True only if the given network has more than 3
         """
         try:
-            resp = self.driver_v2.network_connection.request(
-                "/v2.0/network-ip-availabilities/{}".format(network),
+            url = f"/v2.0/network-ip-availabilities/{net.id}"
+            resp = self.driver_v2.network_connection.request(url)
+            subnets = resp.object.get("network_ip_availability", {}).get(
+                "subnet_ip_availability", []
             )
-        except BaseException as be:
-            logger.info(be)
-            return False
 
-        if resp.status != 200:
-            logger.debug("Failed to gather the network details.")
-            return False
+            for subnet in subnets:
+                _free_ips = subnet.get("total_ips") - subnet.get("used_ips")
 
-        _total = resp.object.get("network_ip_availability", {}).get("total_ips")
-        _used = resp.object.get("network_ip_availability", {}).get("used_ips")
+                if _free_ips > 3:
+                    self.subnet = subnet.get("cidr")
+                    return True
+        except BaseException as be:  # noqa
+            logger.warning(be)
 
-        return (_total - _used) > 3
+        logger.info("%s has no IP addresses to lease.", net.name)
+        return False
 
     def _get_network(self, network: Optional[str] = None) -> OpenStackNetwork:
         """
-        Return the provider network.
+        Return the instance of OpenStackNetwork from the known list or given name.
+
+        A network object is returned only if it contains a free IP address to lease.
```

Comment:
Let's rewrite this text to be more specific:

```
Return the first available OpenStackNetwork with a free IP address to lease.

This method will search a preconfigured list of network names and return the
first one with a free IP address. You can override the list of networks by
specifying a single network's name.
```

Reply:
done.